server/src/main/java/org/spine3/server/validate/FieldValidator.java

```diff
@@ -0,0 +1,199 @@
+import static java.lang.String.format;
+
+/**
+ * Validates a message field according to Spine custom protobuf options and provides validation error messages.
+ *
+ * @param <V> a type of field values
+ * @author Alexander Litus
+ */
+/* package */ abstract class FieldValidator<V> {
+
+    private boolean isFieldInvalid = false;
+    private final List<String> errorMessages = newLinkedList();
+    private final FieldDescriptor fieldDescriptor;
+    private final ImmutableList<V> values;
+
+    /**
+     * Creates a new validator instance.
+     *
+     * @param descriptor a descriptor of the field to validate
+     */
+    protected FieldValidator(FieldDescriptor descriptor, ImmutableList<V> values) {
+        this.fieldDescriptor = descriptor;
+        this.values = values;
+    }
+
+    /**
+     * Validates a message field according to Spine custom protobuf options and sets validation error messages.
+     *
+     * <p>Uses {@link #setIsFieldInvalid(boolean)} and {@link #addErrorMessage(String)} methods.
+     */
+    protected abstract void validate();
+
+    /**
+     * Returns an immutable list of the field values.
+     */
+    @SuppressWarnings("ReturnOfCollectionOrArrayField") // is immutable list
+    protected ImmutableList<V> getValues() {
+        return values;
+    }
+
+    /**
+     * Checks if the field is required and not set.
+     *
+     * <p>If the field is repeated, it must have at least one value set, and all its values must be valid.
+     *
+     * <p>It is required to override {@link #isValueNotSet(Object)} method to use this one.
+     */
+    protected void checkIfRequiredAndNotSet() {
+        final RequiredOption option = getFieldOption(ValidationProto.required);
+        if (!option.getIs()) {
+            return;
+        }
+        if (values.isEmpty()) {
+            setIsFieldInvalid(true);
+            addErrorMessage(option);
+            return;
+        }
+        for (V value : values) {
+            if (isValueNotSet(value)) {
+                setIsFieldInvalid(true);
+                addErrorMessage(option);
+                return; // return because one error message is enough for the "required" option
+            }
+        }
+    }
+
+    /**
+     * Checks if the field value is not set.
+     *
+     * <p>If the field type is {@link Message}, it must be set to a non-default instance;
+     * if it is {@link String} or {@link ByteString}, it must be set to a non-empty string or array.
+     *
+     * <p>The default implementation throws {@link UnsupportedOperationException}.
+     *
+     * @param value a field value to check
+     * @return {@code true} if the field is not set, {@code false} otherwise
+     */
+    protected boolean isValueNotSet(V value) {
+        throw new UnsupportedOperationException("The method is not implemented.");
+    }
+
+    /**
+     * Returns {@code true} if the validated field is invalid, {@code false} otherwise.
+     */
+    protected boolean isFieldInvalid() {
```

Comment:
Why don't have it `isFieldValid` or just `isValid`? It's the field validator after all.